File: frontend/lib/models/household.dart
// models/household.dart

class Household {
  final int id;
  final String name;
  final DateTime createdAt;
  final List<HouseholdMember> members;

  Household({
    required this.id,
    required this.name,
    required this.createdAt,
    this.members = const [],
  });

  factory Household.fromJson(Map<String, dynamic> json) {
    return Household(
      id: json['id'],
      name: json['name'],
      createdAt: DateTime.parse(json['created_at']),
      members: (json['members'] as List<dynamic>?)
              ?.map((member) => HouseholdMember.fromJson(member))
              .toList() ??
          [],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'created_at': createdAt.toIso8601String(),
      'members': members.map((member) => member.toJson()).toList(),
    };
  }
}

class HouseholdSummary {
  final int id;
  final String name;
  final DateTime? createdAt;

  HouseholdSummary({
    required this.id,
    required this.name,
    this.createdAt,
  });

  factory HouseholdSummary.fromJson(Map<String, dynamic> json) {
    return HouseholdSummary(
      id: json['id'],
      name: json['name'],
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      if (createdAt != null) 'created_at': createdAt!.toIso8601String(),
    };
  }
}

class HouseholdMember {
  final int id;
  final String username;
  final String email;
  final String role;
  final DateTime joinedAt;

  HouseholdMember({
    required this.id,
    required this.username,
    required this.email,
    required this.role,
    required this.joinedAt,
  });

  factory HouseholdMember.fromJson(Map<String, dynamic> json) {
    return HouseholdMember(
      id: json['id'],
      username: json['username'],
      email: json['email'],
      role: json['role'] ?? 'member',
      joinedAt: DateTime.parse(json['joined_at']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'username': username,
      'email': email,
      'role': role,
      'joined_at': joinedAt.toIso8601String(),
    };
  }

  bool get isAdmin => role == 'admin';
}

class HouseholdCreate {
  final String name;

  HouseholdCreate({required this.name});

  Map<String, dynamic> toJson() {
    return {'name': name};
  }
}

class HouseholdUpdate {
  final String name;

  HouseholdUpdate({required this.name});

  Map<String, dynamic> toJson() {
    return {'name': name};
  }
}

class HouseholdInvitation {
  final int id;
  final int householdId;
  final int invitedById;
  final int invitedUserId;
  final String status;
  final DateTime createdAt;
  final DateTime? respondedAt;
  final HouseholdSummary? household;
  final UserSummary? invitedBy;
  final UserSummary? invitedUser;

  HouseholdInvitation({
    required this.id,
    required this.householdId,
    required this.invitedById,
    required this.invitedUserId,
    required this.status,
    required this.createdAt,
    this.respondedAt,
    this.household,
    this.invitedBy,
    this.invitedUser,
  });

  factory HouseholdInvitation.fromJson(Map<String, dynamic> json) {
    return HouseholdInvitation(
      id: json['id'],
      householdId: json['household_id'],
      invitedById: json['invited_by_id'],
      invitedUserId: json['invited_user_id'],
      status: json['status'],
      createdAt: DateTime.parse(json['created_at']),
      respondedAt: json['responded_at'] != null
          ? DateTime.parse(json['responded_at'])
          : null,
      household: json['household'] != null
          ? HouseholdSummary.fromJson(json['household'])
          : null,
      invitedBy: json['invited_by'] != null
          ? UserSummary.fromJson(json['invited_by'])
          : null,
      invitedUser: json['invited_user'] != null
          ? UserSummary.fromJson(json['invited_user'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'household_id': householdId,
      'invited_by_id': invitedById,
      'invited_user_id': invitedUserId,
      'status': status,
      'created_at': createdAt.toIso8601String(),
      if (respondedAt != null) 'responded_at': respondedAt!.toIso8601String(),
      if (household != null) 'household': household!.toJson(),
      if (invitedBy != null) 'invited_by': invitedBy!.toJson(),
      if (invitedUser != null) 'invited_user': invitedUser!.toJson(),
    };
  }

  bool get isPending => status == 'pending';
  bool get isAccepted => status == 'accepted';
  bool get isRejected => status == 'rejected';
  bool get isCancelled => status == 'cancelled';
}

class UserSummary {
  final int id;
  final String username;
  final String email;

  UserSummary({
    required this.id,
    required this.username,
    required this.email,
  });

  factory UserSummary.fromJson(Map<String, dynamic> json) {
    return UserSummary(
      id: json['id'],
      username: json['username'],
      email: json['email'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'username': username,
      'email': email,
    };
  }
}

class InvitationCreate {
  final String email;

  InvitationCreate({required this.email});

  Map<String, dynamic> toJson() {
    return {'email': email};
  }
}

File: backend/app/models/household.py
from sqlalchemy import Column, Integer, String, ForeignKey, DateTime, func
from sqlalchemy.orm import relationship
from datetime import datetime

from app.core.database import Base
from .user import user_households

class Household(Base):
    __tablename__ = "households"
    
    id = Column(Integer, primary_key=True, index=True)
    name = Column(String, index=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    
    # Many-to-many relationship with users
    members = relationship(
        "User", 
        secondary=user_households, 
        back_populates="households"
    )

    invitations = relationship("HouseholdInvitation", back_populates="household")
    shopping_lists = relationship("ShoppingList", back_populates="household")

    def get_members_with_roles(self, db_session):
        from sqlalchemy import text
        result = db_session.execute(
            text("""
                SELECT u.id, u.username, u.email, uh.role, uh.joined_at
                FROM users u 
                JOIN user_households uh ON u.id = uh.user_id 
                WHERE uh.household_id = :household_id
            """),
            {"household_id": self.id}
        ).fetchall()
        
        return [
            {
                "id": row[0],
                "username": row[1], 
                "email": row[2],
                "role": row[3],
                "joined_at": row[4]
            }
            for row in result
        ]

class HouseholdInvitation(Base):
    __tablename__ = "household_invitations"
    
    id = Column(Integer, primary_key=True, index=True)
    household_id = Column(Integer, ForeignKey("households.id"), nullable=False)
    invited_by_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    invited_user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    status = Column(String(20), default="pending")  # pending, accepted, rejected, cancelled
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    responded_at = Column(DateTime(timezone=True), nullable=True)
    
    # Relationships
    household = relationship("Household", back_populates="invitations")
    invited_by = relationship("User", foreign_keys=[invited_by_id])
    invited_user = relationship("User", foreign_keys=[invited_user_id])